File: packages/js/src/node/wav.ts
import { gzipSync, gunzipSync } from "node:zlib";
import { encode, decode, scan } from "../core/index.js";
import type { DecodeOptions, DecodeResult, EncodeOptions, EncodeResult, ScanOptions, ScanResult } from "../core/index.js";

export type WavFormat = "pcm16" | "float32";

export interface WavData {
  sampleRate: number;
  channels: number;
  format: WavFormat;
  samples: Float32Array;
}

export interface EncodeWavOptions extends EncodeOptions {
  wavFormat?: WavFormat;
}

export interface DecodeWavOptions extends Omit<DecodeOptions, "samples"> {
  wav: Uint8Array;
}

export interface ScanWavOptions extends Omit<ScanOptions, "samples"> {
  wav: Uint8Array;
}

export interface EncodeWavResult extends EncodeResult {
  wav: Uint8Array;
}

export interface EncodeWavSamplesOptions {
  samples: Float32Array;
  sampleRate: number;
  format?: WavFormat;
}

export interface DecodeWavSamplesOptions {
  wav: Uint8Array;
}

export interface PrependWavOptions extends Omit<EncodeWavOptions, "json"> {
  wav: Uint8Array;
  json: unknown;
  padSeconds?: number;
  prePadSeconds?: number;
  postPadSeconds?: number;
}

export interface PrependWavResult {
  wav: Uint8Array;
  payload: EncodeResult;
  sampleRate: number;
}

export function encodeWav(options: EncodeWavOptions): EncodeWavResult {
  const wavFormat = options.wavFormat ?? "pcm16";
  const result = encode(withNodeGzip(options));
  const wav = encodeWavSamples({ samples: result.samples, sampleRate: result.sampleRate, format: wavFormat });
  return { ...result, wav };
}

export function decodeWav(options: DecodeWavOptions): DecodeResult {
  const { wav, ...rest } = options;
  const data = decodeWavSamples({ wav });
  return decode(withNodeGunzip({ ...rest, samples: data.samples, sampleRate: rest.sampleRate ?? data.sampleRate }));
}

export function scanWav(options: ScanWavOptions): ScanResult[] {
  const { wav, ...rest } = options;
  const data = decodeWavSamples({ wav });
  return scan(withNodeGunzip({ ...rest, samples: data.samples, sampleRate: rest.sampleRate ?? data.sampleRate }));
}

export function prependPayloadToWav(options: PrependWavOptions): PrependWavResult {
  const { wav, json, ...rest } = options;
  const input = decodeWavSamples({ wav });
  const sampleRate = rest.sampleRate ?? input.sampleRate;
  if (sampleRate !== input.sampleRate) {
    throw new Error(
      `Sample rate mismatch: input ${input.sampleRate} Hz, requested ${sampleRate} Hz. Resampling not supported.`
    );
  }

  const payload = encode(withNodeGzip({ ...rest, json, sampleRate }));
  const padSeconds = rest.padSeconds ?? 0.25;
  const prePadSeconds = rest.prePadSeconds ?? padSeconds;
  const postPadSeconds = rest.postPadSeconds ?? padSeconds;
  const prePadSamples = secondsToSamples(sampleRate, prePadSeconds);
  const postPadSamples = secondsToSamples(sampleRate, postPadSeconds);

  const combined = new Float32Array(
    prePadSamples + payload.samples.length + postPadSamples + input.samples.length
  );
  combined.set(payload.samples, prePadSamples);
  combined.set(input.samples, prePadSamples + payload.samples.length + postPadSamples);

  const wavFormat = rest.wavFormat ?? "pcm16";
  const wavOut = encodeWavSamples({ samples: combined, sampleRate, format: wavFormat });
  return { wav: wavOut, payload, sampleRate };
}

export function encodeWavSamples(options: EncodeWavSamplesOptions): Uint8Array {
  const { samples, sampleRate, format = "pcm16" } = options;
  const numChannels = 1;
  const bitsPerSample = format === "float32" ? 32 : 16;
  const bytesPerSample = bitsPerSample / 8;
  const blockAlign = numChannels * bytesPerSample;
  const byteRate = sampleRate * blockAlign;
  const dataSize = samples.length * bytesPerSample;
  const headerSize = 44;
  const buffer = new ArrayBuffer(headerSize + dataSize);
  const view = new DataView(buffer);

  writeString(view, 0, "RIFF");
  view.setUint32(4, 36 + dataSize, true);
  writeString(view, 8, "WAVE");

  writeString(view, 12, "fmt ");
  view.setUint32(16, 16, true);
  view.setUint16(20, format === "float32" ? 3 : 1, true);
  view.setUint16(22, numChannels, true);
  view.setUint32(24, sampleRate, true);
  view.setUint32(28, byteRate, true);
  view.setUint16(32, blockAlign, true);
  view.setUint16(34, bitsPerSample, true);

  writeString(view, 36, "data");
  view.setUint32(40, dataSize, true);

  if (format === "float32") {
    let offset = headerSize;
    for (let i = 0; i < samples.length; i += 1) {
      view.setFloat32(offset, clamp(samples[i]), true);
      offset += 4;
    }
  } else {
    let offset = headerSize;
    for (let i = 0; i < samples.length; i += 1) {
      const value = Math.round(clamp(samples[i]) * 32767);
      view.setInt16(offset, value, true);
      offset += 2;
    }
  }

  return new Uint8Array(buffer);
}

export function decodeWavSamples(options: DecodeWavSamplesOptions): WavData {
  const { wav } = options;
  const view = new DataView(wav.buffer, wav.byteOffset, wav.byteLength);
  if (readString(view, 0, 4) !== "RIFF" || readString(view, 8, 4) !== "WAVE") {
    throw new Error("Invalid WAV header");
  }

  let offset = 12;
  let format: number | null = null;
  let channels = 0;
  let sampleRate = 0;
  let bitsPerSample = 0;
  let dataOffset = 0;
  let dataSize = 0;

  while (offset + 8 <= view.byteLength) {
    const chunkId = readString(view, offset, 4);
    const chunkSize = view.getUint32(offset + 4, true);
    const chunkDataOffset = offset + 8;

    if (chunkId === "fmt ") {
      format = view.getUint16(chunkDataOffset, true);
      channels = view.getUint16(chunkDataOffset + 2, true);
      sampleRate = view.getUint32(chunkDataOffset + 4, true);
      bitsPerSample = view.getUint16(chunkDataOffset + 14, true);
    } else if (chunkId === "data") {
      dataOffset = chunkDataOffset;
      dataSize = chunkSize;
    }

    offset = chunkDataOffset + chunkSize + (chunkSize % 2);
  }

  if (format === null || dataOffset === 0) {
    throw new Error("WAV missing fmt or data chunk");
  }
  if (channels < 1) {
    throw new Error("Invalid WAV channel count");
  }

  const bytesPerSample = bitsPerSample / 8;
  const totalFrames = Math.floor(dataSize / (bytesPerSample * channels));
  const samples = new Float32Array(totalFrames);

  if (format === 1 && bitsPerSample === 16) {
    let frameOffset = dataOffset;
    for (let i = 0; i < totalFrames; i += 1) {
      let sum = 0;
      for (let ch = 0; ch < channels; ch += 1) {
        sum += view.getInt16(frameOffset, true) / 32768;
        frameOffset += 2;
      }
      samples[i] = sum / channels;
    }
  } else if (format === 3 && bitsPerSample === 32) {
    let frameOffset = dataOffset;
    for (let i = 0; i < totalFrames; i += 1) {
      let sum = 0;
      for (let ch = 0; ch < channels; ch += 1) {
        sum += view.getFloat32(frameOffset, true);
        frameOffset += 4;
      }
      samples[i] = sum / channels;
    }
  } else {
    throw new Error(`Unsupported WAV format ${format} with ${bitsPerSample} bits`);
  }

  return {
    sampleRate,
    channels,
    format: format === 3 ? "float32" : "pcm16",
    samples,
  };
}

function clamp(value: number): number {
  if (value > 1) return 1;
  if (value < -1) return -1;
  return value;
}

function writeString(view: DataView, offset: number, text: string): void {
  for (let i = 0; i < text.length; i += 1) {
    view.setUint8(offset + i, text.charCodeAt(i));
  }
}

function readString(view: DataView, offset: number, length: number): string {
  let text = "";
  for (let i = 0; i < length; i += 1) {
    text += String.fromCharCode(view.getUint8(offset + i));
  }
  return text;
}

function withNodeGzip(options: EncodeWavOptions): EncodeWavOptions {
  if (options.gzipCompress) {
    return options;
  }
  return {
    ...options,
    gzipCompress: (data) => new Uint8Array(gzipSync(data)),
  };
}

function withNodeGunzip<T extends DecodeOptions | ScanOptions>(options: T): T {
  if (options.gzipDecompress) {
    return options;
  }
  return {
    ...options,
    gzipDecompress: (data) => new Uint8Array(gunzipSync(data)),
  };
}

function secondsToSamples(sampleRate: number, seconds: number): number {
  if (!Number.isFinite(seconds) || seconds < 0) {
    return 0;
  }
  return Math.round(seconds * sampleRate);
}

File: packages/python/qraudio/io/wav.py
from __future__ import annotations

import struct
from typing import Literal, Optional, Union

from ..decode import decode, scan
from ..encode import encode
from ..profiles import Profile, normalizeProfile
from ..types import DecodeResult, EncodeResult, EncodeWavResult, PrependWavResult, ScanResult, WavData

WavFormat = Literal["pcm16", "float32"]


def encodeWav(
    payload: object,
    *,
    wav_format: WavFormat = "pcm16",
    **encode_options,
) -> EncodeWavResult:
    result = encode(payload, **encode_options)
    wav = encodeWavSamples(result.samples, result.sampleRate, wav_format)
    return EncodeWavResult(
        sampleRate=result.sampleRate,
        profile=result.profile,
        samples=result.samples,
        durationMs=result.durationMs,
        payloadBytes=result.payloadBytes,
        wav=wav,
    )


def decodeWav(
    wav_bytes: bytes,
    *,
    sample_rate: Optional[int] = None,
    profile: Optional[Union[Profile, str]] = None,
    **options,
) -> DecodeResult:
    data = decodeWavSamples(wav_bytes)
    resolved_profile = normalizeProfile(profile) if profile is not None else None
    return decode(
        data.samples,
        sample_rate=sample_rate or data.sampleRate,
        profile=resolved_profile,
        **options,
    )


def scanWav(
    wav_bytes: bytes,
    *,
    sample_rate: Optional[int] = None,
    profile: Optional[Union[Profile, str]] = None,
    **options,
) -> list[ScanResult]:
    data = decodeWavSamples(wav_bytes)
    resolved_profile = normalizeProfile(profile) if profile is not None else None
    return scan(
        data.samples,
        sample_rate=sample_rate or data.sampleRate,
        profile=resolved_profile,
        **options,
    )


def prependPayloadToWav(
    wav_bytes: bytes,
    payload: object,
    *,
    pad_seconds: float = 0.25,
    pre_pad_seconds: Optional[float] = None,
    post_pad_seconds: Optional[float] = None,
    wav_format: WavFormat = "pcm16",
    **encode_options,
) -> PrependWavResult:
    input_data = decodeWavSamples(wav_bytes)
    sample_rate = encode_options.get("sample_rate", input_data.sampleRate)
    if sample_rate != input_data.sampleRate:
        raise ValueError(
            f"Sample rate mismatch: input {input_data.sampleRate} Hz, requested {sample_rate} Hz. Resampling not supported."
        )

    payload_result = encode(payload, sample_rate=sample_rate, **encode_options)
    pre_pad = pad_seconds if pre_pad_seconds is None else pre_pad_seconds
    post_pad = pad_seconds if post_pad_seconds is None else post_pad_seconds

    pre_samples = secondsToSamples(sample_rate, pre_pad)
    post_samples = secondsToSamples(sample_rate, post_pad)

    combined = [0.0] * (
        pre_samples + len(payload_result.samples) + post_samples + len(input_data.samples)
    )
    combined[pre_samples : pre_samples + len(payload_result.samples)] = payload_result.samples
    offset = pre_samples + len(payload_result.samples) + post_samples
    combined[offset : offset + len(input_data.samples)] = input_data.samples

    wav_out = encodeWavSamples(combined, sample_rate, wav_format)
    return PrependWavResult(wav=wav_out, payload=payload_result, sampleRate=sample_rate)


def encodeWavSamples(samples: list[float], sample_rate: int, fmt: WavFormat = "pcm16") -> bytes:
    num_channels = 1
    bits_per_sample = 32 if fmt == "float32" else 16
    bytes_per_sample = bits_per_sample // 8
    block_align = num_channels * bytes_per_sample
    byte_rate = sample_rate * block_align
    data_size = len(samples) * bytes_per_sample
    header_size = 44

    buffer = bytearray(header_size + data_size)
    buffer[0:4] = b"RIFF"
    struct.pack_into("<I", buffer, 4, 36 + data_size)
    buffer[8:12] = b"WAVE"

    buffer[12:16] = b"fmt "
    struct.pack_into("<I", buffer, 16, 16)
    struct.pack_into("<H", buffer, 20, 3 if fmt == "float32" else 1)
    struct.pack_into("<H", buffer, 22, num_channels)
    struct.pack_into("<I", buffer, 24, sample_rate)
    struct.pack_into("<I", buffer, 28, byte_rate)
    struct.pack_into("<H", buffer, 32, block_align)
    struct.pack_into("<H", buffer, 34, bits_per_sample)

    buffer[36:40] = b"data"
    struct.pack_into("<I", buffer, 40, data_size)

    offset = header_size
    if fmt == "float32":
        for sample in samples:
            struct.pack_into("<f", buffer, offset, clamp(sample))
            offset += 4
    else:
        for sample in samples:
            value = int(round(clamp(sample) * 32767))
            struct.pack_into("<h", buffer, offset, value)
            offset += 2

    return bytes(buffer)


def decodeWavSamples(wav_bytes: bytes) -> WavData:
    if len(wav_bytes) < 12:
        raise ValueError("Invalid WAV header")
    if wav_bytes[0:4] != b"RIFF" or wav_bytes[8:12] != b"WAVE":
        raise ValueError("Invalid WAV header")

    offset = 12
    fmt_tag = None
    channels = 0
    sample_rate = 0
    bits_per_sample = 0
    data_offset = 0
    data_size = 0

    while offset + 8 <= len(wav_bytes):
        chunk_id = wav_bytes[offset : offset + 4]
        chunk_size = struct.unpack_from("<I", wav_bytes, offset + 4)[0]
        chunk_data_offset = offset + 8

        if chunk_id == b"fmt ":
            fmt_tag = struct.unpack_from("<H", wav_bytes, chunk_data_offset)[0]
            channels = struct.unpack_from("<H", wav_bytes, chunk_data_offset + 2)[0]
            sample_rate = struct.unpack_from("<I", wav_bytes, chunk_data_offset + 4)[0]
            bits_per_sample = struct.unpack_from("<H", wav_bytes, chunk_data_offset + 14)[0]
        elif chunk_id == b"data":
            data_offset = chunk_data_offset
            data_size = chunk_size

        offset = chunk_data_offset + chunk_size + (chunk_size % 2)

    if fmt_tag is None or data_offset == 0:
        raise ValueError("WAV missing fmt or data chunk")
    if channels < 1:
        raise ValueError("Invalid WAV channel count")

    bytes_per_sample = bits_per_sample // 8
    total_frames = data_size // (bytes_per_sample * channels)
    samples: list[float] = []

    if fmt_tag == 1 and bits_per_sample == 16:
        frame_offset = data_offset
        for _ in range(total_frames):
            total = 0.0
            for _ in range(channels):
                value = struct.unpack_from("<h", wav_bytes, frame_offset)[0]
                total += value / 32768.0
                frame_offset += 2
            samples.append(total / channels)
        fmt = "pcm16"
    elif fmt_tag == 3 and bits_per_sample == 32:
        frame_offset = data_offset
        for _ in range(total_frames):
            total = 0.0
            for _ in range(channels):
                value = struct.unpack_from("<f", wav_bytes, frame_offset)[0]
                total += value
                frame_offset += 4
            samples.append(total / channels)
        fmt = "float32"
    else:
        raise ValueError(f"Unsupported WAV format {fmt_tag} with {bits_per_sample} bits")

    return WavData(sampleRate=sample_rate, channels=channels, format=fmt, samples=samples)


def secondsToSamples(sample_rate: int, seconds: float) -> int:
    return max(1, round(seconds * sample_rate))


def clamp(value: float) -> float:
    if value > 1.0:
        return 1.0
    if value < -1.0:
        return -1.0
    return value
